File: backend/src/crawl/problem.rs
use anyhow::Context;
use itertools::Itertools;
use sqlx::{Acquire, Pool, Postgres};
use std::{collections::BTreeSet, time::Duration};
use tracing::instrument;

use crate::atcoder::{AtCoderClient, AtCoderProblemsClient, Problem};

#[instrument(skip(atcoder_client, problems_client, pool))]
pub async fn crawl_problems<'a>(
    atcoder_client: &'a AtCoderClient,
    problems_client: &'a AtCoderProblemsClient,
    pool: &'a Pool<Postgres>,
    all: bool,
    duration: Duration,
) -> anyhow::Result<()> {
    let mut targets = problems_client.fetch_problems().await?;
    if !all {
        targets = detect_diff(pool, &targets).await?;
    }

    tracing::info!("start to crawl {} problems", targets.len());

    for target in targets.iter() {
        let html = atcoder_client
            .fetch_problem_html(&target.contest_id, &target.id)
            .await?;

        let mut tx = pool.begin().await?;
        insert_problem(&mut tx, target, &html).await?;

        tx.commit().await?;

        tracing::info!("saved problem {} successfully", target.id);
        tokio::time::sleep(duration).await;
    }

    Ok(())
}

async fn detect_diff<'a, A>(db: A, problems: &[Problem]) -> anyhow::Result<Vec<Problem>>
where
    A: Acquire<'a, Database = Postgres>,
{
    let mut conn = db.acquire().await?;

    let rows: Vec<(String,)> = sqlx::query_as(r#"SELECT "problem_id" FROM "problems";"#)
        .fetch_all(&mut *conn)
        .await
        .with_context(|| "fetch problem ids from database")?;
    let exists = BTreeSet::from_iter(rows.iter().map(|r| r.0.clone()));

    let result = problems
        .iter()
        .filter(|p| !exists.contains(&p.id))
        .cloned()
        .collect_vec();

    Ok(result)
}

#[instrument(skip(db, problem, html))]
async fn insert_problem<'a, A>(db: A, problem: &Problem, html: &str) -> anyhow::Result<u64>
where
    A: Acquire<'a, Database = Postgres>,
{
    let sql = r#"
INSERT INTO
    "problems" (
        "problem_id",
        "contest_id",
        "problem_index",
        "name",
        "title",
        "url",
        "html",
        "updated_at"
    )
VALUES
    ($1, $2, $3, $4, $5, $6, $7, NOW())
ON CONFLICT ("problem_id") DO
UPDATE
SET
    "contest_id" = EXCLUDED."contest_id",
    "problem_index" = EXCLUDED."problem_index",
    "name" = EXCLUDED."name",
    "title" = EXCLUDED."title",
    "url" = EXCLUDED."url",
    "html" = EXCLUDED."html",
    "updated_at" = NOW();
    "#;

    let mut conn = db.acquire().await?;
    let res = sqlx::query(sql)
        .bind(&problem.id)
        .bind(&problem.contest_id)
        .bind(&problem.problem_index)
        .bind(&problem.name)
        .bind(&problem.title)
        .bind(&problem.url())
        .bind(html)
        .execute(&mut *conn)
        .await
        .with_context(|| "execute insert problem query")?;

    Ok(res.rows_affected())
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::testutil::{create_container, create_pool_from_container};
    use sqlx::Row;
    use testcontainers::runners::AsyncRunner;

    #[tokio::test]
    async fn test_insert_problem() {
        let container = create_container().unwrap().start().await.unwrap();
        let pool = create_pool_from_container(&container).await.unwrap();

        let problem = Problem {
            id: String::from("abc001_a"),
            contest_id: String::from("abc001"),
            problem_index: String::from("A"),
            title: String::from("title"),
            name: String::from("A. title"),
        };
        let html = "html";

        let count = insert_problem(&pool, &problem, &html).await.unwrap();
        assert_eq!(count, 1);

        let sql = r#"SELECT * FROM "problems""#;
        let rows = sqlx::query(sql).fetch_all(&pool).await.unwrap();
        assert_eq!(rows.len(), 1);
        let row = &rows[0];

        assert_eq!(
            problem.id,
            row.try_get::<String, &str>("problem_id").unwrap()
        );
        assert_eq!(
            problem.contest_id,
            row.try_get::<String, &str>("contest_id").unwrap()
        );
        assert_eq!(
            problem.problem_index,
            row.try_get::<String, &str>("problem_index").unwrap()
        );
        assert_eq!(problem.name, row.try_get::<String, &str>("name").unwrap());
        assert_eq!(problem.title, row.try_get::<String, &str>("title").unwrap());
        assert_eq!("html", row.try_get::<String, &str>("html").unwrap());
        assert_eq!(
            "https://atcoder.jp/contests/abc001/tasks/abc001_a",
            row.try_get::<String, &str>("url").unwrap()
        );
    }

    #[tokio::test]
    async fn test_detect_diff() {
        let container = create_container().unwrap().start().await.unwrap();
        let pool = create_pool_from_container(&container).await.unwrap();

        let sql = r#"
INSERT INTO "problems" ("problem_id", "contest_id", "problem_index", "name", "title", "url", "html")
VALUES
    ('abc001_a', 'abc001', 'A', 'test problem 1', 'A. test problem 1', 'url', 'html');
        "#;
        let res = sqlx::query(sql).execute(&pool).await.unwrap();
        assert_eq!(res.rows_affected(), 1);

        let problems = vec![
            Problem {
                id: String::from("abc001_a"),
                contest_id: String::from("abc001"),
                problem_index: String::from("A"),
                name: String::from("test problem 1"),
                title: String::from("A. test problem 1"),
            },
            Problem {
                id: String::from("abc001_b"),
                contest_id: String::from("abc001"),
                problem_index: String::from("B"),
                name: String::from("test problem 2"),
                title: String::from("B. test problem 2"),
            },
        ];

        let diff = detect_diff(&pool, &problems).await.unwrap();
        let want = vec![problems[1].clone()];
        assert_eq!(diff, want);
    }
}

File: backend/crawl/problem.go
package crawl

import (
	"bytes"
	"context"
	"database/sql"
	"fjnkt98/atcodersearch/atcoder"
	"fjnkt98/atcodersearch/repository"
	"fmt"
	"log/slog"
	"maps"
	"slices"
	"strings"
	"time"

	mapset "github.com/deckarep/golang-set/v2"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/jackc/pgx/v5/stdlib"
	"github.com/tdewolff/minify"
	"github.com/tdewolff/minify/html"
	"github.com/uptrace/bun"
	"github.com/uptrace/bun/dialect/pgdialect"
)

type ProblemCrawler struct {
	atcoderClient  atcoder.AtCoderClient
	problemsClient atcoder.AtCoderProblemsClient
	pool           *pgxpool.Pool
	duration       time.Duration
	all            bool
	minifier       *minify.M
}

func NewProblemCrawler(
	atcoderClient atcoder.AtCoderClient,
	problemsClient atcoder.AtCoderProblemsClient,
	pool *pgxpool.Pool,
	duration time.Duration,
	all bool,
) *ProblemCrawler {
	m := minify.New()
	m.AddFunc("text/html", html.Minify)

	return &ProblemCrawler{
		atcoderClient:  atcoderClient,
		problemsClient: problemsClient,
		pool:           pool,
		minifier:       m,
		duration:       duration,
		all:            all,
	}
}

func (c *ProblemCrawler) CrawlContests(ctx context.Context) error {
	slog.LogAttrs(ctx, slog.LevelInfo, "start to crawl contests.")

	contests, err := c.problemsClient.FetchContests(ctx)
	if err != nil {
		return fmt.Errorf("fetch contests: %w", err)
	}

	count, err := SaveContests(ctx, c.pool, contests)
	if err != nil {
		return fmt.Errorf("save contests: %w", err)
	}

	slog.LogAttrs(ctx, slog.LevelInfo, "finish to crawl contests.", slog.Int64("count", count))
	return nil
}

func (c *ProblemCrawler) CrawlDifficulties(ctx context.Context) error {
	slog.LogAttrs(ctx, slog.LevelInfo, "start to crawl difficulties.")

	difficulties, err := c.problemsClient.FetchDifficulties(ctx)
	if err != nil {
		return fmt.Errorf("fetch difficulties: %w", err)
	}

	count, err := SaveDifficulties(ctx, c.pool, difficulties)
	if err != nil {
		return fmt.Errorf("save difficulties: %w", err)
	}

	slog.LogAttrs(ctx, slog.LevelInfo, "finish to crawl difficulties.", slog.Int64("count", count))
	return nil
}

func (c *ProblemCrawler) CrawlProblems(ctx context.Context) error {
	targets, err := c.problemsClient.FetchProblems(ctx)
	if err != nil {
		return fmt.Errorf("fetch problems: %w", err)
	}

	if !c.all {
		targets, err = DetectDiff(ctx, c.pool, targets)
		if err != nil {
			return fmt.Errorf("detect diff: %w", err)
		}
	}
	slog.LogAttrs(ctx, slog.LevelInfo, "start to crawl problems.", slog.Int("targets", len(targets)))

	save := func(ctx context.Context, p atcoder.Problem) (int64, error) {
		html, err := c.atcoderClient.FetchProblemHTML(ctx, p.ContestID, p.ID)
		if err != nil {
			return 0, fmt.Errorf("fetch problem html: %w", err)
		}

		tx, err := c.pool.Begin(ctx)
		if err != nil {
			return 0, fmt.Errorf("begin transaction: %w", err)
		}
		defer tx.Rollback(ctx)

		q := repository.New(c.pool).WithTx(tx)

		var buf bytes.Buffer
		if err := c.minifier.Minify("text/html", &buf, strings.NewReader(html)); err != nil {
			return 0, fmt.Errorf("minify html: %w", err)
		}

		res, err := q.InsertProblem(ctx, repository.InsertProblemParams{
			ProblemID:    p.ID,
			ContestID:    p.ContestID,
			ProblemIndex: p.ProblemIndex,
			Name:         p.Name,
			Title:        p.Title,
			Url:          p.URL(),
			Html:         buf.String(),
		})
		if err != nil {
			return 0, fmt.Errorf("insert problem: %w", err)
		}

		if err := tx.Commit(ctx); err != nil {
			return 0, fmt.Errorf("commit transaction: %w", err)
		}

		return res.RowsAffected(), nil
	}

	var count int64 = 0
	for _, p := range targets {
		if c, err := save(ctx, p); err != nil {
			return err
		} else {
			count += c
		}
		slog.LogAttrs(ctx, slog.LevelInfo, "save problem", slog.String("problemID", p.ID))
		time.Sleep(c.duration)
	}

	slog.LogAttrs(ctx, slog.LevelInfo, "finish to crawl problems.", slog.Int64("count", count))

	return nil
}

func DetectDiff(ctx context.Context, pool *pgxpool.Pool, problems []atcoder.Problem) ([]atcoder.Problem, error) {
	q := repository.New(pool)
	ids, err := q.FetchProblemIDs(ctx)
	if err != nil {
		return nil, fmt.Errorf("fetch problem ids: %w", err)
	}

	exists := mapset.NewSet(ids...)

	diff := make([]atcoder.Problem, 0, len(problems))
	for _, p := range problems {
		if !exists.Contains(p.ID) {
			diff = append(diff, p)
		}
	}

	return diff, nil
}

type Contest struct {
	bun.BaseModel    `bun:"table:contests,alias:c"`
	ContestID        string `bun:"contest_id"`
	StartEpochSecond int64  `bun:"start_epoch_second"`
	DurationSecond   int64  `bun:"duration_second"`
	Title            string `bun:"title"`
	RateChange       string `bun:"rate_change"`
	Category         string `bun:"category"`
}

func NewContests(contests []atcoder.Contest) []Contest {
	result := make([]Contest, len(contests))

	for i, c := range contests {
		result[i] = Contest{
			ContestID:        c.ID,
			StartEpochSecond: c.StartEpochSecond,
			DurationSecond:   c.DurationSecond,
			Title:            c.Title,
			RateChange:       c.RateChange,
			Category:         c.Categorize(),
		}
	}

	return result
}

func SaveContests(ctx context.Context, pool *pgxpool.Pool, contests []atcoder.Contest) (int64, error) {
	if len(contests) == 0 {
		return 0, nil
	}

	tx, err := bun.NewDB(stdlib.OpenDBFromPool(pool), pgdialect.New()).BeginTx(ctx, &sql.TxOptions{})
	if err != nil {
		return 0, fmt.Errorf("begin transaction: %w", err)
	}
	defer tx.Rollback()

	var count int64 = 0

	for chunk := range slices.Chunk(NewContests(contests), 1000) {
		res, err := tx.NewInsert().
			Model(&chunk).
			On("CONFLICT (contest_id) DO UPDATE").
			Set("contest_id = EXCLUDED.contest_id").
			Set("start_epoch_second = EXCLUDED.start_epoch_second").
			Set("duration_second = EXCLUDED.duration_second").
			Set("title = EXCLUDED.title").
			Set("rate_change = EXCLUDED.rate_change").
			Set("category = EXCLUDED.category").
			Set("updated_at = NOW()").
			Exec(ctx)
		if err != nil {
			return 0, fmt.Errorf("insert: %w", err)
		}

		if c, err := res.RowsAffected(); err != nil {
			return 0, fmt.Errorf("rows affected: %w", err)
		} else {
			count += c
		}
	}

	if err := tx.Commit(); err != nil {
		return 0, fmt.Errorf("commit transaction: %w", err)
	}

	return count, nil
}

type Difficulty struct {
	bun.BaseModel    `bun:"table:difficulties,alias:d"`
	ProblemID        string   `bun:"problem_id"`
	Slope            *float64 `bun:"slope"`
	Intercept        *float64 `bun:"intercept"`
	Variance         *float64 `bun:"variance"`
	Difficulty       *int64   `bun:"difficulty"`
	Discrimination   *float64 `bun:"discrimination"`
	IrtLoglikelihood *float64 `bun:"irt_loglikelihood"`
	IrtUsers         *float64 `bun:"irt_users"`
	IsExperimental   *bool    `bun:"is_experimental"`
}

func NewDifficulties(difficulties map[string]atcoder.Difficulty) []Difficulty {
	result := make([]Difficulty, len(difficulties))

	for i, problemID := range slices.Sorted(maps.Keys(difficulties)) {
		d := difficulties[problemID]

		result[i] = Difficulty{
			ProblemID:        problemID,
			Slope:            d.Slope,
			Intercept:        d.Intercept,
			Variance:         d.Variance,
			Difficulty:       d.Difficulty,
			Discrimination:   d.Discrimination,
			IrtLoglikelihood: d.IrtLoglikelihood,
			IrtUsers:         d.IrtUsers,
			IsExperimental:   d.IsExperimental,
		}
	}

	return result
}

func SaveDifficulties(ctx context.Context, pool *pgxpool.Pool, difficulties map[string]atcoder.Difficulty) (int64, error) {
	if len(difficulties) == 0 {
		return 0, nil
	}

	tx, err := bun.NewDB(stdlib.OpenDBFromPool(pool), pgdialect.New()).BeginTx(ctx, &sql.TxOptions{})
	if err != nil {
		return 0, fmt.Errorf("begin transaction: %w", err)
	}
	defer tx.Rollback()

	var count int64 = 0

	for chunk := range slices.Chunk(NewDifficulties(difficulties), 1000) {
		res, err := tx.NewInsert().
			Model(&chunk).
			On("CONFLICT (problem_id) DO UPDATE").
			Set("problem_id = EXCLUDED.problem_id").
			Set("slope = EXCLUDED.slope").
			Set("intercept = EXCLUDED.intercept").
			Set("variance = EXCLUDED.variance").
			Set("difficulty = EXCLUDED.difficulty").
			Set("discrimination = EXCLUDED.discrimination").
			Set("irt_loglikelihood = EXCLUDED.irt_loglikelihood").
			Set("irt_users = EXCLUDED.irt_users").
			Set("is_experimental = EXCLUDED.is_experimental").
			Set("updated_at = NOW()").
			Exec(ctx)
		if err != nil {
			return 0, fmt.Errorf("insert: %w", err)
		}

		if c, err := res.RowsAffected(); err != nil {
			return 0, fmt.Errorf("rows affected: %w", err)
		} else {
			count += c
		}
	}

	if err := tx.Commit(); err != nil {
		return 0, fmt.Errorf("commit transaction: %w", err)
	}

	return count, nil
}
